crypto/fipsmodule/kdf/sskdf.c

```diff
@@ -0,0 +1,314 @@
+                                        const uint8_t *secret,
+                                        size_t secret_len, const uint8_t *info,
+                                        size_t info_len) {
+  if (!ctx || !ctx->data || !out || !counter || !secret) {
+    return 0;
+  }
+
+  SSKDF_VARIANT_DIGEST_CTX *variant_ctx = (SSKDF_VARIANT_DIGEST_CTX *)ctx->data;
+
+  if (!variant_ctx->ctx || !variant_ctx->digest) {
+    return 0;
+  }
+
+  if (!EVP_MD_CTX_reset(variant_ctx->ctx) ||
+      !EVP_DigestInit_ex(variant_ctx->ctx, variant_ctx->digest, NULL) ||
+      !EVP_DigestUpdate(variant_ctx->ctx, &counter[0], 4) ||
+      !EVP_DigestUpdate(variant_ctx->ctx, secret, secret_len) ||
+      !EVP_DigestUpdate(variant_ctx->ctx, info, info_len) ||
+      !EVP_DigestFinal(variant_ctx->ctx, out, NULL)) {
+    return 0;
+  }
+
+  return 1;
+}
+
+static size_t sskdf_variant_hmac_size(SSKDF_VARIANT_CTX *ctx) {
+  if (!ctx || !ctx->data) {
+    return 0;
+  }
+  SSKDF_VARIANT_HMAC_CTX *variant_ctx = (SSKDF_VARIANT_HMAC_CTX *)ctx->data;
+  if (!variant_ctx) {
+    return 0;
+  }
+  return HMAC_size(variant_ctx->ctx);
+}
+
+static int sskdf_variant_hmac_compute(SSKDF_VARIANT_CTX *ctx, uint8_t *out,
+                                      size_t out_len, const uint8_t counter[4],
+                                      const uint8_t *secret, size_t secret_len,
+                                      const uint8_t *info, size_t info_len) {
+  if (!ctx || !ctx->data || !out || !counter || !secret) {
+    return 0;
+  }
+
+  SSKDF_VARIANT_HMAC_CTX *variant_ctx = (SSKDF_VARIANT_HMAC_CTX *)ctx->data;
+
+  if (!variant_ctx->ctx) {
+    return 0;
+  }
+
+  if (!HMAC_Init_ex(variant_ctx->ctx, NULL, 0, NULL, NULL) ||
+      !HMAC_Update(variant_ctx->ctx, &counter[0], 4) ||
+      !HMAC_Update(variant_ctx->ctx, secret, secret_len) ||
+      !HMAC_Update(variant_ctx->ctx, info, info_len) ||
+      !HMAC_Final(variant_ctx->ctx, out, NULL)) {
+    return 0;
+  }
+
+  return 1;
+}
+
+
+DEFINE_METHOD_FUNCTION(SSKDF_VARIANT, SSKDF_VARIANT_DIGEST) {
+  memset(out, 0, sizeof(SSKDF_VARIANT));
+  out->size = sskdf_variant_digest_size;
+  out->compute = sskdf_variant_digest_compute;
+}
+
+
+DEFINE_METHOD_FUNCTION(SSKDF_VARIANT, SSKDF_VARIANT_HMAC) {
+  memset(out, 0, sizeof(SSKDF_VARIANT));
+  out->size = sskdf_variant_hmac_size;
+  out->compute = sskdf_variant_hmac_compute;
+};
+
+static int SSKDF(const SSKDF_VARIANT *variant, SSKDF_VARIANT_CTX *ctx,
+                 uint8_t *out_key, size_t out_len, const uint8_t *secret,
+                 size_t secret_len, const uint8_t *info, size_t info_len) {
+  int ret = 0;
+
+  if (!ctx) {
+    abort();
+  }
+
+  if (!out_key || out_len == 0 || !secret || secret_len == 0) {
+    goto err;
+  }
+
+  size_t h_len = variant->size(ctx);
+  if (!h_len) {
+    goto err;
+  }
+  assert(h_len <= EVP_MAX_MD_SIZE);
+
+  uint64_t n = (out_len + h_len - 1) / h_len;
+
+  if (out_len + h_len < out_len || n > UINT32_MAX) {
```

Comment:
nit:  you already check above if h_len == 0 so its impossible for out_len + h_len < out_len.